README.md

```diff
@@ -18,36 +18,203 @@ testing first appeard in Erlang's proprietary QuickCheck. The
+```
+
+Finally, we have to explain how to generate and shrink actions.
+
+```haskell
+generator :: Model Symbolic -> Gen (Untyped Action)
+generator (Model m)
+  | null m    = pure (Untyped New)
+  | otherwise = frequency
+      [ (1, pure (Untyped New))
+      , (8, Untyped .    Read  <$> elements (map fst m))
+      , (8, Untyped <$> (Write <$> elements (map fst m) <*> arbitrary))
+      , (8, Untyped .    Inc   <$> elements (map fst m))
+      ]
+
+shrinker :: Action v resp -> [Action v resp]
+shrinker (Write ref i) = [ Write ref i' | i' <- shrink i ]
+shrinker _             = []
+```
 
-┌──────────────────────┐
-│ New                  │
-│                 ⟶ $0 │
-└──────────────────────┘
-            │ ┌────────┐
-            │ │ Inc $0 │
-┌─────────┐ │ │        │
-│ Inc $0  │ │ │        │
-│         │ │ │   ⟶ () │
-│         │ │ └────────┘
-│    ⟶ () │ │
-└─────────┘ │
-┌─────────┐ │
-│ Read $0 │ │
-│     ⟶ 1 │ │
-└─────────┘ │
+We can now define a sequential and a parallel property as follows.
+
+```haskell
+prop_references :: Problem -> Property
+prop_references prb = forAllProgram
+  generator
+  shrinker
+  precondition
+  transition
+  initModel $ \prog ->
+    runAndCheckProgram
+      precondition
+      transition
+      postcondition
+      initModel
+      (semantics prb)
+      ioProperty
+      prog
+
+prop_referencesParallel :: Problem -> Property
+prop_referencesParallel prb = forAllParallelProgram
+  generator
+  shrinker
+  precondition
+  transition
+  initModel $ \parallel ->
+    runParallelProgram (semantics prb) parallel $ \hist ->
+      checkParallelProgram
+        transition
+        postcondition
+        initModel
+        parallel
+        hist
+```
 
+If we run the sequential property without introducing any problems to the
+semantics function, i.e. `quickCheck (prop_references None)`, then the property
+passes. If we however introduce the bug problem, then it will fail with the
+minimal counterexample:
 
 ```
+> quickCheck (prop_references Bug)
+*** Failed! Falsifiable (after 16 tests and 4 shrinks):
+[New (Var 0),Write (Var 0) 5 (Var 2),Read (Var 0) (Var 3)]
+Just 5 /= Just 6
+```
+
+Running the sequential property with the race condition problem will not uncover
+the race condition, but the parallel property will!
+
+```
+> quickCheck (prop_referencesParallel RaceCondition)
+*** Failed! (after 8 tests and 6 shrinks):
+
+Couldn't linearise:
+
+┌────────────────────────────────┐
+│ Var 0 ← New                    │
+│                       ⟶ Opaque │
+└────────────────────────────────┘
+┌─────────────┐ │
+│ Inc (Var 0) │ │
+│             │ │ ┌──────────────┐
+│             │ │ │ Inc (Var 0)  │
+│        ⟶ () │ │ │              │
+└─────────────┘ │ │              │
+                │ │         ⟶ () │
+                │ └──────────────┘
+                │ ┌──────────────┐
+                │ │ Read (Var 0) │
+                │ │          ⟶ 1 │
+                │ └──────────────┘
+Just 2 /= Just 1
+```
```

Comment:
Quick comment how to read this diagram, that explains that this program creates a new `Var 0`, and then in parallel (concurrently) does two `Inc`, but somehow only reads a `1` afterwards. Which if one looks in the semantics can see that we have a race condition.